boot/bootstate20.go

```diff
@@ -37,20 +41,49 @@ func newBootState20(typ snap.Type) bootState {
 	}
 }
 
+// newEarlyBootState20 is like newBootState20, but it initializes the
+// bootState with the provided bootloader and/or modeenv - mainly to be used
+// with the initramfs logic where we load the modeenv and bootloader slightly
+// differently than normal.
+func newEarlyBootState20(typ snap.Type) earlyBootState {
+	switch typ {
+	case snap.TypeBase:
+		bs := &bootState20Base{}
+		bs.earlyBoot = true
+		return bs
+	case snap.TypeKernel:
+		bs := &bootState20Kernel{earlyBoot: true}
+		return bs
+	case snap.TypeSnapd:
+		bs := &bootState20Snapd{}
+		bs.earlyBoot = true
+		return bs
+	default:
+		panic(fmt.Sprintf("cannot make a bootState20 for snap type %q", typ))
+	}
+}
+
 //
 // modeenv methods
 //
 
 type bootState20Modeenv struct {
 	modeenv *Modeenv
+
+	// this just controls where to try to find the modeenv
+	earlyBoot bool
 }
 
 func (bsm *bootState20Modeenv) loadModeenv() error {
 	// don't read modeenv multiple times
 	if bsm.modeenv != nil {
 		return nil
 	}
-	modeenv, err := ReadModeenv("")
+	dir := ""
+	if bsm.earlyBoot {
+		dir = filepath.Join(dirs.EarlyBootUbuntuData, "system-data")
```

Comment:
I really don't like that boot learns the precise layouts used by initramfs

Reply:
hmm I did have an earlier variant of this which carried the bootloader options and modeenv dir to read from in something like `newBootState20WithOpts(...)`, perhaps that would simplify this, but if your concern is that nothing in the boot package should know this then we would muddle the interface quite a bit with InitramfsRunModeSnapsToMount needing to take in various things like bootloader options or modeenv dirs, etc.

Reply:
maybe this would be simpler if we have something like bootloader options but for modeenv that would at least look less weird to carry around